### PR TITLE
chore(wallet): remove fiatBalance from Redux and use base amount units

### PR DESCRIPTION
**Pull request by onyb**
**Summary of changes:**
- Add native asset price to `transactionSpotPrices` field, and remove `fiatPrice`.
- Do not update balances when refreshing prices, and vice-versa.
- Do not store computed `fiatBalance` in Redux, and instead compute it from `assetBalance` and `transactionSpotPrices` using the `usePricing` custom hook.
- Do not perform amount-based computations on floating point numbers. Always use `BigNumber` for addition, multiplication, etc.
- Always represent amounts/balances in Redux using base units (ex, wei for Ethereum).

Related to https://github.com/brave/brave-browser/issues/20264.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Demo

TBA

### CODE REVIEW

File: components/brave_wallet_ui/utils/bn-utils.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ */
+export function normalizeNumericValue (value: string): string {
+  return value === ''
+    ? ''
+    : new BigNumber(value).toFixed(0)
+}
+
+export function isNumericValueGreaterThan (a: string, b: string): boolean {
+  return new BigNumber(a).isGreaterThan(b)
+}
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/utils/bn-utils.ts:50]: Should we have some tests for these?

File: components/brave_wallet_ui/common/hooks/pricing.ts [hunk 2]
```diff
@@ -6,11 +6,21 @@
 import * as React from 'react'
 
 import { BraveWallet } from '../../constants/types'
+import { formatFiatBalance } from '../../utils/format-balances'
 
 export default function usePricing (spotPrices: BraveWallet.AssetPrice[]) {
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/common/hooks/pricing.ts:11]: Should we add a test for this hook?

**Review by Douglashdaniel — CHANGES_REQUESTED** [review]
Nice refactor!
Left a few comments.

Additionally while testing this locally, it looks like something has re-introduced showing `0.00` balances before the actual value is returned. 

https://user-images.githubusercontent.com/40611140/148588909-a6f7bf14-c8f3-4743-aaa9-12a8043f99cc.mov

**Comment by onyb** [comment]
@Douglashdaniel Good catch on unknown balances being displayed as `0.00`. I noticed we had the same bug with transaction/gas-fee values too, so I fixed that as well. Here's a screen recording.

https://user-images.githubusercontent.com/3684187/148723514-8845ba45-d302-4531-a388-1effb2f49790.mov

File: components/brave_wallet_ui/common/hooks/pricing.ts (hunk 2 above)
Comment (onyb) [reply components/brave_wallet_ui/common/hooks/pricing.ts:11]: Done!

File: components/brave_wallet_ui/utils/bn-utils.ts (hunk 1 above)
Comment (onyb) [reply components/brave_wallet_ui/utils/bn-utils.ts:50]: Done

File: components/brave_wallet_ui/common/hooks/swap.ts [hunk 3]
```diff
@@ -410,13 +410,15 @@ export default function useSwap (
     }
 
     if (swapValidationError === 'insufficientAllowance' && allowance) {
-      const fromAssetBalanceWeiHex = toWeiHex(fromAssetBalance, fromAsset.asset.decimals)
+      const allowanceHex = !fromAssetBalance
+        ? toWeiHex(fromAmount, fromAsset.asset.decimals)
+        : toHex(fromAssetBalance)
```
Comment (onyb) [inline components/brave_wallet_ui/common/hooks/swap.ts:415]: If we're unable to fetch the balance (ex, if asset is in swap options list, but not in visible options list), we set the allowance to the value of the swap transaction.

**Review by Douglashdaniel — APPROVED** [review]
Looks good!
Nice work!

**Pull request merged**